sv.c

```diff
@@ -4771,6 +4776,272 @@ Perl_sv_setsv_flags(pTHX_ SV *dsv, SV* ssv, const I32 flags)
+                   Have to copy the string.  */
+                sv_grow_fresh(dsv, cur + 1);
+
+                Move(SvPVX_const(ssv),SvPVX(dsv),cur,char);
+                SvCUR_set(dsv, cur);
+                *SvEND(dsv) = '\0';
+            }
+        SvFLAGS(dsv) |= sflags & SVf_UTF8;
+}
+
+/* A helper for sv_mortalcopy_flags and newSVsv_flags, more efficient
+ * for their use cases compared to Perl_sv_setsv_flags. */
+PERL_STATIC_INLINE SV *
+S_sv_freshcopy_flags(pTHX_ SV* ssv, const I32 flags)
+{
+    /* Note: If a PVIV/PVNV/PVMG is only IOK, NOK, ROK, it is _mostly_
+     * possible to create just a headless SV to store that value.
+     * Some parts of core (Perl_amagic_call in gv.c specifically) do
+     * assume - and possibly CPAN might - that SvTYPE(dsv) == SvTYPE(ssv)
+     * though, which is why the code below does not try that type
+     * simplification. Perhaps this might be worth revisiting in the future.
+     * -- April 2025. */
+    assert(ssv);
+    svtype stype = SvTYPE(ssv);
+
+#if NVSIZE <= IVSIZE
+    ASSUME(stype > SVt_NV);
+#else
+    ASSUME(stype > SVt_IV);
+#endif
+    ASSUME(stype <= SVt_PVMG);
+    SV *dsv = newSV_type(stype);
+    U32 sflags   = SvFLAGS(ssv);
+
+    ASSUME(!(flags & SV_GMAGIC)); /* Caller should have sorted this. */
+    /* Note: Perl_sv_setsv_flags essentially ignores magic, except for
+     * taint and vstring magic. */
+
+    /* The switch cases are in the order that gcov of make/make test suggests. */
+    switch(sflags & (SVp_IOK|SVp_NOK|SVf_ROK|SVp_POK|SVf_FAKE|SVppv_STATIC)) {
+        case SVp_IOK:
+            SvIV_set(dsv, SvIVX(ssv));
+            SvFLAGS(dsv) |= sflags & (SVf_IOK|SVp_IOK|SVf_IVisUV);
+            break;
+        case SVp_POK:
+            if (UNLIKELY(SvTYPE(ssv) == SVt_INVLIST)) {
+               invlist_clone(ssv, dsv);
+               return dsv;
+            }
+            S_sv_freshcopy_POK(aTHX_ dsv, ssv, flags);
+        {
+            const char *vstr_pv;
+            STRLEN vstr_len;
+            if ((vstr_pv = SvVSTRING(ssv, vstr_len))) {
+                sv_magic(dsv, NULL, PERL_MAGIC_vstring, vstr_pv, vstr_len);
+                SvRMAGICAL_on(dsv);
+            }
+        }
+            break;
+        case SVp_POK|SVp_IOK:
+            S_sv_freshcopy_POK(aTHX_ dsv, ssv, flags);
+        {
+            const char *vstr_pv;
+            STRLEN vstr_len;
+            if ((vstr_pv = SvVSTRING(ssv, vstr_len))) {
+                sv_magic(dsv, NULL, PERL_MAGIC_vstring, vstr_pv, vstr_len);
+                SvRMAGICAL_on(dsv);
+            }
+        }
+            SvIV_set(dsv, SvIVX(ssv));
+            SvFLAGS(dsv) |= (sflags & (SVf_IOK|SVp_IOK|SVf_IVisUV));
+            if ((sflags & SVf_IOK) && !(sflags & SVf_POK)) {
+                /* Source was SVf_IOK|SVp_IOK|SVp_POK but not SVf_POK, meaning
+                   a value set as an integer and later stringified. So mark
+                   destination the same: */
+                SvFLAGS(dsv) &= ~SVf_POK;
```

Comment:
What code above in this func, already wrote to  dsv SvFLAGS , in such a way that public flag not private flag, SVf_POK bit is turned on in dsv right now? Im confused.

Reply:
I'll double check later, but IIRC `S_sv_freshcopy_POK` sets `SVf_POK` on `dsv`. (This was copy & pasted right out of `Perl_sv_setsv_flags` where the comments mention that the flag is turned on to help with COW. I should also double check whether it actually does help with COW, or if that's some bitrot.)

So the purpose of this snippet - also from `Perl_sv_setsv_flags` IIRC - is to flip off `SVf_POK` if the source SV didn't actually have that flag turned on.

Reply:
`S_sv_freshcopy_POK` does indeed unconditionally set `SVf_POK` on `dsv`. That is behaviour taken from `Perl_sv_setsv_flags`.

 I don't see how that "_facilitates the COW check_" as stated in the comment in `Perl_sv_setsv_flags` though. I'll look into it further.

Reply:
Looks like setting `SVf_POK` on `dsv` when it wasn't on `ssv` is avoidable.

That should remove the need for the above toggle, meaning that some of the cases probably now can be written to fall through.